demos/mqtt/subscription-manager/mqtt_subscription_manager.c

```diff
@@ -0,0 +1,441 @@
+                       callbackRecordList[ listIndex ].topicFilterLength,
+                       callbackRecordList[ listIndex ].pTopicFilter,
+                       pPublishInfo->topicNameLength,
+                       pPublishInfo->pTopicName ) );
+
+            /* Invoke the callback associated with the record as the topics match. */
+            callbackRecordList[ listIndex ].callback( pContext, pPublishInfo );
+        }
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+bool SubscriptionManager_RegisterCallback( const char * pTopicFilter,
+                                           uint16_t topicFilterLength,
+                                           SubscriptionManager_Callback_t callback )
+{
+    assert( pTopicFilter != NULL );
+    assert( topicFilterLength != 0 );
+
+    bool recordAdded = false;
+    size_t availableIndex = 0u;
+    bool recordExists = false;
+
+    /* Search for an available spot in the list to store the record */
+    while( ( availableIndex < MAX_SUBSCRIPTION_CALLBACK_RECORDS ) &&
+           ( callbackRecordList[ availableIndex ].pTopicFilter != NULL ) )
+    {
+        /* Check if a record for the topic filter already exists in the registry. */
+        if( ( callbackRecordList[ availableIndex ].topicFilterLength == topicFilterLength ) &&
+            ( strncmp( pTopicFilter, callbackRecordList[ availableIndex ].pTopicFilter, topicFilterLength )
+              == 0 ) )
+        {
+            recordExists = true;
+        }
+
+        availableIndex++;
+    }
+
+    if( ( availableIndex == MAX_SUBSCRIPTION_CALLBACK_RECORDS ) )
+    {
+        /* The record list is full. */
+        LogError( ( "Unable to register callback: Registry list is full: TopicFilter=%.*s",
+                    topicFilterLength,
+                    pTopicFilter ) );
+    }
+    else if( recordExists == true )
+    {
+        /* The record for the topic filter already exists. */
+        LogError( ( "Failed to register callback: Record for topic filter already exists: TopicFilter=%.*s",
+                    topicFilterLength,
+                    pTopicFilter ) );
+    }
+    else
+    {
+        /* Should the topic string be copied? */
+        callbackRecordList[ availableIndex ].pTopicFilter = pTopicFilter;
+        callbackRecordList[ availableIndex ].topicFilterLength = topicFilterLength;
+        callbackRecordList[ availableIndex ].callback = callback;
+
+        recordAdded = true;
+
+        LogDebug( ( "Added callback to registry: TopicFilter=%.*s",
+                    topicFilterLength,
+                    pTopicFilter ) );
+    }
+
+    return recordAdded;
+}
+
+/*-----------------------------------------------------------*/
+
+void SubscriptionManager_RemoveCallback( const char * pTopicFilter,
+                                         uint16_t topicFilterLength )
+{
+    size_t matchingRecordIndex = 0u;
+    bool recordFound = false;
+    SubscriptionManager_Record_t * pRecord = NULL;
+
```

Comment:
Should probably `assert( pTopicFilter != NULL );`

Reply:
Added